### PR TITLE
#462: Adding a build action to build the Dafny parser

### PR DESCRIPTION
**Pull request by davidcok**

### CODE REVIEW

File: .github/workflows/gradle.yml [hunk 1]
```diff
@@ -19,4 +19,6 @@ jobs:
         with:
           java-version: 1.8
       - name: Build with Gradle
-        run: ./Source/DafnyRuntime/DafnyRuntimeJava/gradlew build
+        run: ./Source/DafnyRuntime/DafnyRuntimeJava/gradlew -p ./Source/DafnyRuntime/DafnyRuntimeJava build
```
Comment (robin-aws) [inline .github/workflows/gradle.yml:22]: I'm not 100% sure but I suspect building "test" will invoke "build" automatically, so you can probably remove this step.

File: .github/workflows/msbuild.yml [hunk 2]
```diff
@@ -47,9 +47,19 @@ jobs:
...
+      run: make -C dafny/Source/Dafny -f Makefile.Linux
     - name: Build Dafny
       run: msbuild dafny/Source/Dafny.sln
     - uses: actions/setup-python@v1
+    - name: Set up JDK 1.8
+      uses: actions/setup-java@v1
+      with:
+        java-version: 1.8
+    - name: Build DafnyRuntime.jar with Gradle
+      run: dafny/Source/DafnyRuntime/DafnyRuntimeJava/gradlew -p dafny/Source/DafnyRuntime/DafnyRuntimeJava build
```
Comment (robin-aws) [inline .github/workflows/msbuild.yml:60]: Same comment here, also it depends on whether `copyJarToBinaries` declares its dependencies correctly

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+${DIR}/Binaries/DafnyRuntime.jar:
+	(cd ${DIR}/Source/DafnyRuntime/DafnyRuntimeJava; ./gradlew build && gradle copyJarToBinaries)
+
+refman:
+	make -C ${DIR}/docs/DafnyReferenceManual
+
+clean:
+	(cd ${DIR} ; rm -f Source/Dafny/bin/Debug/* Binaries/dafny.exe )
+	make -C ${DIR}/Source/Dafny clean
+	(cd ${DIR} ; rm -f Binaries/DafnyRuntime.jar )
```
Comment (robin-aws) [inline Makefile:29]: This should probably invoke a clean target using gradle instead, so you also clean up class files and any other junk.

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+runtime: ${DIR}/Binaries/DafnyRuntime.jar
+
+${DIR}/Binaries/DafnyRuntime.jar:
+	(cd ${DIR}/Source/DafnyRuntime/DafnyRuntimeJava; ./gradlew build && gradle copyJarToBinaries)
+
+refman:
+	make -C ${DIR}/docs/DafnyReferenceManual
+
+clean:
+	(cd ${DIR} ; rm -f Source/Dafny/bin/Debug/* Binaries/dafny.exe )
```
Comment (robin-aws) [inline Makefile:27]: This should delegate to msbuild instead to clean IMO

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+	make -C ${DIR}/Source/Dafny clean
+	(cd ${DIR} ; rm -f Binaries/DafnyRuntime.jar )
+	(cd ${DIR} ; rm -f docs/DafnyReferenceManual/DafnyRef.pdf )
```
Comment (robin-aws) [inline Makefile:30]: `make -C docs/DafnyReferenceManual clean` instead?

File: .github/workflows/gradle.yml (hunk 1 above)
Comment (davidcok) [reply .github/workflows/gradle.yml]: Checked and it does. Edited.

File: .github/workflows/msbuild.yml (hunk 2 above)
Comment (davidcok) [reply .github/workflows/msbuild.yml]: Added the dependency and edited correspondingly.

File: Makefile (hunk 4 above)
Comment (davidcok) [reply Makefile]: Done

File: Makefile (hunk 5 above)
Comment (davidcok) [reply Makefile]: Done

File: Makefile (hunk 3 above)
Comment (davidcok) [reply Makefile]: Done

File: Source/DafnyRuntime/DafnyRuntimeJava/build.gradle [hunk 6]
```diff
@@ -26,8 +26,16 @@ test {
 }
 
 task copyJarToBinaries {
-    copy {
+    dependsOn build
```
Comment (robin-aws) [inline Source/DafnyRuntime/DafnyRuntimeJava/build.gradle:29]: Nice!

File: .github/workflows/msbuild.yml [hunk 7]
```diff
@@ -47,9 +47,17 @@ jobs:
         cp -r z3* dafny/Binaries/z3
     - name: Nuget Restore Dafny
       run: nuget restore dafny/Source/Dafny.sln
+    - name: Build Dafny Parser
```
Comment (robin-aws) [inline .github/workflows/msbuild.yml:50]: Make sure you add this to INSTALL.md and any other place we document building Dafny (the wiki has at least one page too)

File: .github/workflows/msbuild.yml [hunk 8]
```diff
@@ -47,9 +47,17 @@ jobs:
...
+      run: make -C dafny/Source/Dafny -f Makefile.Linux
     - name: Build Dafny
       run: msbuild dafny/Source/Dafny.sln
     - uses: actions/setup-python@v1
+    - name: Set up JDK 1.8
+      uses: actions/setup-java@v1
+      with:
+        java-version: 1.8
+    - name: Build and Copy DafnyRuntime.jar with Gradle
+      run: dafny/Source/DafnyRuntime/DafnyRuntimeJava/gradlew -p dafny/Source/DafnyRuntime/DafnyRuntimeJava copyJarToBinaries
```
Comment (robin-aws) [inline .github/workflows/msbuild.yml:60]: Same here (possible in an "if you are using Java" section)

Comment (davidcok) [reply .github/workflows/msbuild.yml:60]: Ditto

File: .github/workflows/msbuild.yml (hunk 7 above)
Comment (davidcok) [reply .github/workflows/msbuild.yml:50]: Added comments, add a separate issue to update the complete INSTALL instructions

**Pull request merged**